keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,766 @@
+  - Cons:
+    - Not aligned with `Finalizers`
+    - Bad scalability (Size of the field and number of API calls required)
+
+#### Race of removing/adding liens
+
+A cluster-admin would be racing to update the object to remove the lien and delete the object before a namespace editor is able to place the lien back. As a result, a namespace admin or editor can create an object that a cluster-admin or namespace lifecycle controller cannot delete, which shouldn't be allowed.
+
+To mitigate the risk, a new `NotBlockOnLien` API option in `DeleteOptions` to force delete a resource with liens will be added.
+
+TODO: Update below description depending on the decision.
+
+```
+Only a limited set of users and groups are allowed to specify the option, therefore deletion request is blocked in api-server,if invalid users request deletion with the option.
+The default users allowed are `system:controller:namespace-controller` and default groups allowed are `system:masters`.
+The lists of allowed users and allowed groups can be modified through new kube-apiserver's command line options, `--allowed-not-block-on-lien-users` and `--allowed-not-block-on-lien-groups`.
+In addition, a new `--not-block-on-lien` kubectl command line option to specify the API option is also added.
+
+There was an alternative idea to mitigate the risk by making liens field a subresource to allow defining a separate RBAC.
+By doing so, a cluster-admin will be able to set the RBAC for the liens to deny update by namespace admin or editor.
+However, a command line option is chosen as a solution, because liens are introduced as an advisory feature to help prevent accidental deletion, not a hard locking feature which sets privileges around who can or cannot lock / delete.
+
+As an advisory feature, anyone should be able to apply a lien, and anyone should be able to ignore them.
+Therefore, a command line option is better suited to this feature.
+```
+
+#### Subresource
+
+There is a use case to disallow a controller to write to its CRD except for the status subresource, but allow it to update finalizers/liens.
+To allow this use case, it is required to provide a feature to set a separate RBAC to update finalizers/liens.
+On the other hand, finalizers field is currently not subresourced.
+Therefore, it needs to provide this feature in a compatible way for finalizers, and it would be better to implement liens field in the same way to finalizers.
+
+To solve this issue, there is [an idea](https://github.com/kubernetes/enhancements/pull/2840#discussion_r723464012) to add a new nob for just patching only a specific field via an endpoint of a subresource, without modifying the field itself.
+In this way, a feature to set a separate RBAC can be implemented later without making liens field itself a subresource.
+Therefore, liens field isn't defined as a subresource and providing a separate RBAC for liens is done in a separate KEP/PR along with finalizers.
```

Comment:
I think subresource is a beta requirement

Reply:
I would expect the API shape to be completed while in alpha, not beta

Reply:
I meant "to enter beta" - are you saying we should hold this from entering alpha?

Reply:
> I meant "to enter beta" - are you saying we should hold this from entering alpha?

not necessarily, I just wouldn't want the first time we see the shape of the subresource API / interactions to be in the release that promoted to beta.

Reply:
to clarify: I think the subresource aspect is one of the more important and tricky aspects to get right, so I'd really encourage figuring that out as part of the initial alpha implementation, rather than shipping an alpha to get feedback on an API we know doesn't resemble the final shape